datatypes/read-value-id.go

```diff
@@ -0,0 +1,213 @@
+package datatypes
+
+import "encoding/binary"
+
+// IntegerID is a UInt32 that is used as an identifier, such as a handle.
+// All values, except for 0, are valid.
+//
+// Specification: Part 4, 7.14
+type IntegerID uint32
+
+// Identifiers assigned to Attributes.
+//
+// Specification: Part 6, A.1
+const (
+	IntegerIDNodeID IntegerID = iota + 1
+	IntegerIDNodeClass
+	IntegerIDBrowseName
+	IntegerIDDisplayName
+	IntegerIDDescription
+	IntegerIDWriteMask
+	IntegerIDUserWriteMask
+	IntegerIDIsAbstract
+	IntegerIDSymmetric
+	IntegerIDInverseName
+	IntegerIDContainsNoLoops
+	IntegerIDEventNotifier
+	IntegerIDValue
+	IntegerIDDataType
+	IntegerIDValueRank
+	IntegerIDArrayDimensions
+	IntegerIDAccessLevel
+	IntegerIDUserAccessLevel
+	IntegerIDMinimumSamplingInterval
+	IntegerIDHistorizing
+	IntegerIDExecutable
+	IntegerIDUserExecutable
+	IntegerIDDataTypeDefinition
+	IntegerIDRolePermissions
+	IntegerIDUserRolePermissions
+	IntegerIDAccessRestrictions
+	IntegerIDAccessLevelEx
+)
+
+// ReadValueID is an identifier for an item to read or to monitor.
+//
+// Specification: Part 4, 7.24
+type ReadValueID struct {
+	NodeID
+	AttributeID  IntegerID
+	IndexRange   *String
+	DataEncoding *QualifiedName
+}
+
+// DecodeReadValueID decodes given bytes into ReadValueID.
+func DecodeReadValueID(b []byte) (*ReadValueID, error) {
+	r := &ReadValueID{}
+	if err := r.DecodeFromBytes(b); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+// DecodeFromBytes decodes given bytes into OPC UA ReadValueID.
+func (r *ReadValueID) DecodeFromBytes(b []byte) error {
+	nodeID, err := DecodeNodeID(b)
+	if err != nil {
+		return err
+	}
+	r.NodeID = nodeID
+	offset := r.NodeID.Len()
+
+	// attribute id
+	r.AttributeID = IntegerID(binary.LittleEndian.Uint32(b[offset:]))
+	offset += 4
+
+	// index range
+	r.IndexRange = &String{}
+	if err := r.IndexRange.DecodeFromBytes(b[offset:]); err != nil {
+		return err
+	}
+	offset += r.IndexRange.Len()
+
+	// data encoding
+	r.DataEncoding = &QualifiedName{}
+	if err := r.DataEncoding.DecodeFromBytes(b[offset:]); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Serialize serializes ReadValueID into bytes.
+func (r *ReadValueID) Serialize() ([]byte, error) {
+	b := make([]byte, r.Len())
+	if err := r.SerializeTo(b); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+// SerializeTo serializes ReadValueID into bytes.
+func (r *ReadValueID) SerializeTo(b []byte) error {
+	// node id
+	if err := r.NodeID.SerializeTo(b); err != nil {
+		return err
+	}
+	offset := r.NodeID.Len()
+
+	// attribute id
+	binary.LittleEndian.PutUint32(b[offset:offset+4], uint32(r.AttributeID))
+	offset += 4
+
+	// index range
+	if err := r.IndexRange.SerializeTo(b[offset:]); err != nil {
```

Comment:
To avoid panic I always put `if xxx != nil {` block here even the spec says the field is mandatory. This may cause generating malformed byte stream, but the purpose is to be as flexible as possible (imagine fuzz-testing some OPC UA server with malformed packets).
Let me know if you have better idea :smiley: